### PR TITLE
Add a warning for disallowed email domains

### LINKED ISSUE #29660: Add a warning when admin add a user with email address not listed in `EMAIL_DOMAIN_ALLOWLIST`
### Feature Description

https://github.com/go-gitea/gitea/pull/29605#issuecomment-1980105119
@Zettat123
avoid forget doing this, as the PR was closed

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by Zettat123**
Resolve #29660

Follow #29522 and #29609

Add a warning for disallowed email domains when admins manually add/edit users.

Thanks @yp05327 for the [comment](https://github.com/go-gitea/gitea/pull/29605#issuecomment-1980105119)

image

### CODE REVIEW

**Comment by lunny** [comment]
related to #29660?

**Comment by yp05327** [comment]
My suggestion:
`The user email domain %s is not listed in EMAIL_DOMAIN_ALLOWLIST. Please ensure your operation is expected.`

ps: does other options effect it? e.g. something like email domain not allowed.

**Comment by Zettat123** [comment]
> ps: does other options effect it?

Yes. `EMAIL_DOMAIN_BLOCKLIST` also affects.

Maybe `The user email domain %s conflicts with EMAIL_DOMAIN_ALLOWLIST or EMAIL_DOMAIN_BLOCKLIST. Please ensure your operation is expected.` ?

**Comment by Zettat123** [comment]
> ps: is it possible to add warning in the API?

Not sure how to add the warning to the API. Maybe something like this?

https://github.com/go-gitea/gitea/blob/9dc8a6336edddca0f3f90392bbc398f4ceaeaf13/routers/api/v1/repo/repo.go#L1190-L1192

File: routers/web/admin/users.go [hunk 1]
```diff
@@ -202,6 +202,11 @@ func NewUserPost(ctx *context.Context) {
 		}
 		return
 	}
+
+	if !user_model.IsEmailDomainAllowed(u.Email) {
+		ctx.Flash.Warning(ctx.Tr("form.email_domain_is_not_allowed"))
```
Comment (delvh) [inline routers/web/admin/users.go]: Preferably, the email (domain) would be part of the warning.

File: routers/web/admin/users.go [hunk 2]
```diff
@@ -425,6 +430,9 @@ func EditUserPost(ctx *context.Context) {
 			}
 			return
 		}
+		if !user_model.IsEmailDomainAllowed(form.Email) {
+			ctx.Flash.Warning(ctx.Tr("form.email_domain_is_not_allowed"))
```
Comment (delvh) [inline routers/web/admin/users.go]: See above

File: routers/web/admin/users.go (hunk 1 above)
Comment (Zettat123) [reply routers/web/admin/users.go]: Done in 39f4efd4b7cdcd9f4d1ca651d59df70411b0d650

File: routers/web/admin/users.go (hunk 2 above)
Comment (Zettat123) [reply routers/web/admin/users.go]: Done in 39f4efd4b7cdcd9f4d1ca651d59df70411b0d650

**Review by denyskon — APPROVED** [review]

**Comment by yp05327** [comment]
> > ps: is it possible to add warning in the API?
> 
> Not sure how to add the warning to the API. Maybe something like this?
> 
> https://github.com/go-gitea/gitea/blob/9dc8a6336edddca0f3f90392bbc398f4ceaeaf13/routers/api/v1/repo/repo.go#L1190-L1192

Searched a lot, and normally there are three solutions:
- some Apps are using http status code 299 as a warning http status code. 
- there's a `Warning` header, but it is deprecated:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Warning
- add warnings in JSON body. It is impossible for Gitea, as we are using http status code as the result, but not adding `status` in JSON body.

It seems that there's no normal rules.

**Comment by Zettat123** [comment]
> some Apps are using http status code 299 as a warning http status code.

Looks like `299` is not an official code. Since Gitea is already using `X-Gitea-Warning` to indicate warning messages, I think we can also use it for now

**Comment by yp05327** [comment]
> Since Gitea is already using X-Gitea-Warning to indicate warning messages

Thanks for your notification. I didn't notice that.

**Review by yp05327 — APPROVED** [review]

**Pull request merged**